ICE: `escaping bound vars in predicate`

<!--
[31mICE[0m: Rustc ./a.rs '-Zunstable-options --edition=2024 -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: compiler/rustc_hir_typeck/src/typeck_root_ctxt.rs:106:13: escaping bound vars in predicate Obligation(predicate=Binder { value: ProjectionPredicate(AliasTerm { args: [?0t], def_id: DefId(2:8783 ~ core[fdc7]::iter::traits::iterator::Iterator::Item), .. }, Term::Ty(Alias(Opaque, AliasTy { args: ['a/#0, '^2_0.Named(DefId(0:12 ~ a[f61b]::size_of_val::'_), "'_"), 'a/#0], def_id: DefId(0:9 ~ a[f61b]::size_of_val::{opaque#0}::{opaque#0}), .. }))), bound_vars: [] }, depth=0)', 'error: internal compiler error: compiler/rustc_hir_typeck/src/typeck_root_ctxt.rs:106:13: escaping bound vars in predicate Obligation(predicate=Binder { value: ProjectionPredicate(AliasTerm { args: [?0t], def_id: DefId(2:8783 ~ core[fdc7]::iter::traits::iterator::Iterator::Item), .. }, Term::Ty(Alias(Opaque, AliasTy { args: ['a/#0, '^2_0.Named(DefId(0:12 ~ a[f61b]::size_of_val::'_), "'_"), 'a/#0], def_id: DefId(0:9 ~ a[f61b]::size_of_val::{opaque#0}::{opaque#0}), .. }))), bound_vars: [] }, depth=0)'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
fn size_of_val<'a>(x: &mut ()) -> impl Iterator<Item = impl Sized> + use<> {
    x
}


````

original:
````rust
fn lifetime_in_bounds<'a>(x: &'tr ()) -> impl Into<&'a ()> + use<> { x }
//~^ ERROR `impl Trait` captures lifetime parameter, but it is not mentioned in `use<...>` precise captures list

fn size_of_val<'a>(x: &mut ()) -> impl Iterator<Item = impl Sized> + use<> { x }
//~^ ERROR hidden type for `impl Sized` captures lifetime that does not appear in bounds

fn main() {
            iter_raw -= 1;
        }

````

Version information
````
rustc 1.84.0-nightly (9ce3675b4 2024-10-15)
binary: rustc
commit-hash: 9ce3675b438aae22ef0c6147cde2003a418ab722
commit-date: 2024-10-15
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zunstable-options --edition=2024`

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.jJVes92ppekh/rustc_testrunner_tmpdir_reporting.5Srfjlvyeixq/mvce.rs:3:2
  |
3 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.jJVes92ppekh/rustc_testrunner_tmpdir_reporting.5Srfjlvyeixq/mvce.rs`

error: `impl Trait` captures lifetime parameter, but it is not mentioned in `use<...>` precise captures list
 --> /tmp/icemaker_global_tempdir.jJVes92ppekh/rustc_testrunner_tmpdir_reporting.5Srfjlvyeixq/mvce.rs:1:35
  |
1 | fn size_of_val<'a>(x: &mut ()) -> impl Iterator<Item = impl Sized> + use<> {
  |                --                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime captured due to being mentioned in the bounds of the `impl Trait`
  |                |
  |                this lifetime parameter is captured

error: internal compiler error: compiler/rustc_hir_typeck/src/typeck_root_ctxt.rs:106:13: escaping bound vars in predicate Obligation(predicate=Binder { value: ProjectionPredicate(AliasTerm { args: [?0t], def_id: DefId(2:8783 ~ core[fdc7]::iter::traits::iterator::Iterator::Item), .. }, Term::Ty(Alias(Opaque, AliasTy { args: ['a/#0, '^2_0.Named(DefId(0:7 ~ mvce[11b1]::size_of_val::'_), "'_"), 'a/#0], def_id: DefId(0:6 ~ mvce[11b1]::size_of_val::{opaque#0}::{opaque#0}), .. }))), bound_vars: [] }, depth=0)
 --> /tmp/icemaker_global_tempdir.jJVes92ppekh/rustc_testrunner_tmpdir_reporting.5Srfjlvyeixq/mvce.rs:1:35
  |
1 | fn size_of_val<'a>(x: &mut ()) -> impl Iterator<Item = impl Sized> + use<> {
  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at compiler/rustc_hir_typeck/src/typeck_root_ctxt.rs:106:13:
Box<dyn Any>
stack backtrace:
   0:     0x7382d2c76daa - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h525ca73b44ff3961
   1:     0x7382d340368a - core::fmt::write::h4e0877406968c391
   2:     0x7382d47a9111 - std::io::Write::write_fmt::h2bd0eeb556cd4584
   3:     0x7382d2c76c02 - std::sys::backtrace::BacktraceLock::print::h43c0f0423e8dc1a8
   4:     0x7382d2c790e6 - std::panicking::default_hook::{{closure}}::h15d7535f314f3156
   5:     0x7382d2c78f30 - std::panicking::default_hook::hdfe0b9b278f6a2d4
   6:     0x7382d1cfa68f - std[e857fc88149b8bb7]::panicking::update_hook::<alloc[14f3b7019767b1ae]::boxed::Box<rustc_driver_impl[12a88fbe870fdaff]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x7382d2c797f8 - std::panicking::rust_panic_with_hook::h80214b0734ae1ad1
   8:     0x7382d1d345b1 - std[e857fc88149b8bb7]::panicking::begin_panic::<rustc_errors[9857333b1cca239b]::ExplicitBug>::{closure#0}
   9:     0x7382d1d275e6 - std[e857fc88149b8bb7]::sys::backtrace::__rust_end_short_backtrace::<std[e857fc88149b8bb7]::panicking::begin_panic<rustc_errors[9857333b1cca239b]::ExplicitBug>::{closure#0}, !>
  10:     0x7382d1d22e20 - std[e857fc88149b8bb7]::panicking::begin_panic::<rustc_errors[9857333b1cca239b]::ExplicitBug>
  11:     0x7382d1d3e061 - <rustc_errors[9857333b1cca239b]::diagnostic::BugAbort as rustc_errors[9857333b1cca239b]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7382d099d3a0 - <rustc_errors[9857333b1cca239b]::DiagCtxtHandle>::span_bug::<rustc_span[e1ff7264637c5d63]::span_encoding::Span, alloc[14f3b7019767b1ae]::string::String>
  13:     0x7382d1f93647 - rustc_middle[d8a02af54bfa0dde]::util::bug::opt_span_bug_fmt::<rustc_span[e1ff7264637c5d63]::span_encoding::Span>::{closure#0}
  14:     0x7382d1f9670a - rustc_middle[d8a02af54bfa0dde]::ty::context::tls::with_opt::<rustc_middle[d8a02af54bfa0dde]::util::bug::opt_span_bug_fmt<rustc_span[e1ff7264637c5d63]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7382d1f44c3b - rustc_middle[d8a02af54bfa0dde]::ty::context::tls::with_context_opt::<rustc_middle[d8a02af54bfa0dde]::ty::context::tls::with_opt<rustc_middle[d8a02af54bfa0dde]::util::bug::opt_span_bug_fmt<rustc_span[e1ff7264637c5d63]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7382d008fd97 - rustc_middle[d8a02af54bfa0dde]::util::bug::span_bug_fmt::<rustc_span[e1ff7264637c5d63]::span_encoding::Span>
  17:     0x7382d37c3fe9 - <rustc_hir_typeck[4162f95293710ac]::typeck_root_ctxt::TypeckRootCtxt>::register_predicate
  18:     0x7382d378ec7b - rustc_hir_typeck[4162f95293710ac]::check::check_fn
  19:     0x7382d3783393 - rustc_hir_typeck[4162f95293710ac]::typeck
  20:     0x7382d3782d0f - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>
  21:     0x7382d39044fd - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::VecCache<rustc_span[e1ff7264637c5d63]::def_id::LocalDefId, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  22:     0x7382d390324d - rustc_query_impl[6f041a62200b4f1f]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  23:     0x7382d3d1207f - rustc_middle[d8a02af54bfa0dde]::query::plumbing::query_get_at::<rustc_query_system[6bc689bd08eb534a]::query::caches::VecCache<rustc_span[e1ff7264637c5d63]::def_id::LocalDefId, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>>
  24:     0x7382d43a26eb - rustc_hir_analysis[d00687496dda5c67]::collect::type_of::type_of_opaque
  25:     0x7382d43a2625 - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>
  26:     0x7382d34304ae - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::DefIdCache<rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  27:     0x7382d45d6cf6 - rustc_query_impl[6f041a62200b4f1f]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7382d3cfc202 - rustc_middle[d8a02af54bfa0dde]::query::plumbing::query_get_at::<rustc_query_system[6bc689bd08eb534a]::query::caches::DefIdCache<rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>>
  29:     0x7382d04d5785 - rustc_hir_analysis[d00687496dda5c67]::collect::type_of::type_of
  30:     0x7382d34317b0 - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>
  31:     0x7382d34304ae - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::DefIdCache<rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  32:     0x7382d3430067 - rustc_query_impl[6f041a62200b4f1f]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7382d3cfc202 - rustc_middle[d8a02af54bfa0dde]::query::plumbing::query_get_at::<rustc_query_system[6bc689bd08eb534a]::query::caches::DefIdCache<rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 8usize]>>>
  34:     0x7382d43acf90 - rustc_hir_analysis[d00687496dda5c67]::check::check::check_item_type
  35:     0x7382d0383600 - rustc_hir_analysis[d00687496dda5c67]::check::wfcheck::check_well_formed
  36:     0x7382d3bb103d - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>
  37:     0x7382d3bb0794 - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::VecCache<rustc_span[e1ff7264637c5d63]::def_id::LocalDefId, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  38:     0x7382d3bb0410 - rustc_query_impl[6f041a62200b4f1f]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7382d3bb10bd - rustc_middle[d8a02af54bfa0dde]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[6bc689bd08eb534a]::query::caches::VecCache<rustc_span[e1ff7264637c5d63]::def_id::LocalDefId, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>, ()>
  40:     0x7382d3bb167f - rustc_hir_analysis[d00687496dda5c67]::check::wfcheck::check_mod_type_wf
  41:     0x7382d3bb10e5 - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>
  42:     0x7382d42d3afb - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::DefaultCache<rustc_span[e1ff7264637c5d63]::def_id::LocalModDefId, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  43:     0x7382d42d38ad - rustc_query_impl[6f041a62200b4f1f]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  44:     0x7382d3900abb - rustc_hir_analysis[d00687496dda5c67]::check_crate
  45:     0x7382d38fd817 - rustc_interface[bbaab2745c9b053b]::passes::run_required_analyses
  46:     0x7382d429de1e - rustc_interface[bbaab2745c9b053b]::passes::analysis
  47:     0x7382d429ddf1 - rustc_query_impl[6f041a62200b4f1f]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6f041a62200b4f1f]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>
  48:     0x7382d42e81ae - rustc_query_system[6bc689bd08eb534a]::query::plumbing::try_execute_query::<rustc_query_impl[6f041a62200b4f1f]::DynamicConfig<rustc_query_system[6bc689bd08eb534a]::query::caches::SingleCache<rustc_middle[d8a02af54bfa0dde]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6f041a62200b4f1f]::plumbing::QueryCtxt, false>
  49:     0x7382d42e7e8f - rustc_query_impl[6f041a62200b4f1f]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  50:     0x7382d41aab5e - rustc_interface[bbaab2745c9b053b]::interface::run_compiler::<core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>, rustc_driver_impl[12a88fbe870fdaff]::run_compiler::{closure#0}>::{closure#1}
  51:     0x7382d423fe54 - std[e857fc88149b8bb7]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[bbaab2745c9b053b]::util::run_in_thread_with_globals<rustc_interface[bbaab2745c9b053b]::util::run_in_thread_pool_with_globals<rustc_interface[bbaab2745c9b053b]::interface::run_compiler<core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>, rustc_driver_impl[12a88fbe870fdaff]::run_compiler::{closure#0}>::{closure#1}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>::{closure#0}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>
  52:     0x7382d4240268 - <<std[e857fc88149b8bb7]::thread::Builder>::spawn_unchecked_<rustc_interface[bbaab2745c9b053b]::util::run_in_thread_with_globals<rustc_interface[bbaab2745c9b053b]::util::run_in_thread_pool_with_globals<rustc_interface[bbaab2745c9b053b]::interface::run_compiler<core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>, rustc_driver_impl[12a88fbe870fdaff]::run_compiler::{closure#0}>::{closure#1}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>::{closure#0}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[fdc7858511f3567b]::result::Result<(), rustc_span[e1ff7264637c5d63]::ErrorGuaranteed>>::{closure#1} as core[fdc7858511f3567b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  53:     0x7382d4240d2b - std::sys::pal::unix::thread::Thread::new::thread_start::hd84c48c307791fe8
  54:     0x7382ce4a339d - <unknown>
  55:     0x7382ce52849c - <unknown>
  56:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.84.0-nightly (9ce3675b4 2024-10-15) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unstable-options -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck] type-checking `size_of_val`
#1 [type_of_opaque] computing type of opaque `size_of_val::{opaque#0}`
end of query stack
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
error: internal compiler error: compiler/rustc_hir_typeck/src/typeck_root_ctxt.rs:106:13: escaping bound vars in predicate Obligation(predicate=Binder { value: ProjectionPredicate(AliasTerm { args: [?0t], def_id: DefId(2:8783 ~ core[fdc7]::iter::traits::iterator::Iterator::Item), .. }, Term::Ty(Alias(Opaque, AliasTy { args: ['a/#0, '^2_0.Named(DefId(0:7 ~ mvce[11b1]::size_of_val::'_), "'_"), 'a/#0], def_id: DefId(0:6 ~ mvce[11b1]::size_of_val::{opaque#0}::{opaque#0}), .. }))), bound_vars: [] }, depth=0)
#0 [typeck] type-checking `size_of_val`
#1 [type_of_opaque] computing type of opaque `size_of_val::{opaque#0}`

-->

--- a/compiler/rustc_ast_lowering/src/item.rs
+++ b/compiler/rustc_ast_lowering/src/item.rs
@@ -1574,7 +1574,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
             .collect();
 
         // Introduce extra lifetimes if late resolution tells us to.
-        let extra_lifetimes = self.resolver.take_extra_lifetime_params(parent_node_id);
+        let extra_lifetimes = self.resolver.extra_lifetime_params(parent_node_id);
         params.extend(extra_lifetimes.into_iter().filter_map(|(ident, node_id, res)| {
             self.lifetime_res_to_generic_param(
                 ident,

--- a/compiler/rustc_ast_lowering/src/lib.rs
+++ b/compiler/rustc_ast_lowering/src/lib.rs
@@ -268,8 +268,8 @@ impl ResolverAstLowering {
     ///
     /// The extra lifetimes that appear from the parenthesized `Fn`-trait desugaring
     /// should appear at the enclosing `PolyTraitRef`.
-    fn take_extra_lifetime_params(&mut self, id: NodeId) -> Vec<(Ident, NodeId, LifetimeRes)> {
-        self.extra_lifetime_params_map.remove(&id).unwrap_or_default()
+    fn extra_lifetime_params(&mut self, id: NodeId) -> Vec<(Ident, NodeId, LifetimeRes)> {
+        self.extra_lifetime_params_map.get(&id).cloned().unwrap_or_default()
     }
 }
 
@@ -885,7 +885,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         let mut generic_params: Vec<_> = self
             .lower_generic_params_mut(generic_params, hir::GenericParamSource::Binder)
             .collect();
-        let extra_lifetimes = self.resolver.take_extra_lifetime_params(binder);
+        let extra_lifetimes = self.resolver.extra_lifetime_params(binder);
         debug!(?extra_lifetimes);
         generic_params.extend(extra_lifetimes.into_iter().filter_map(|(ident, node_id, res)| {
             self.lifetime_res_to_generic_param(ident, node_id, res, hir::GenericParamSource::Binder)
@@ -1495,62 +1495,25 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         // frequently opened issues show.
         let opaque_ty_span = self.mark_span_with_reason(DesugaringKind::OpaqueTy, span, None);
 
-        let captured_lifetimes_to_duplicate = if let Some(args) =
-            // We only look for one `use<...>` syntax since we syntactially reject more than one.
-            bounds.iter().find_map(
-                |bound| match bound {
-                    ast::GenericBound::Use(a, _) => Some(a),
-                    _ => None,
-                },
-            ) {
-            // We'll actually validate these later on; all we need is the list of
-            // lifetimes to duplicate during this portion of lowering.
-            args.iter()
-                .filter_map(|arg| match arg {
-                    PreciseCapturingArg::Lifetime(lt) => Some(*lt),
-                    PreciseCapturingArg::Arg(..) => None,
-                })
-                // Add in all the lifetimes mentioned in the bounds. We will error
-                // them out later, but capturing them here is important to make sure
-                // they actually get resolved in resolve_bound_vars.
-                .chain(lifetime_collector::lifetimes_in_bounds(self.resolver, bounds))
-                .collect()
-        } else {
-            match origin {
-                hir::OpaqueTyOrigin::TyAlias { .. } => {
-                    // type alias impl trait and associated type position impl trait were
-                    // decided to capture all in-scope lifetimes, which we collect for
-                    // all opaques during resolution.
-                    self.resolver
-                        .take_extra_lifetime_params(opaque_ty_node_id)
-                        .into_iter()
-                        .map(|(ident, id, _)| Lifetime { id, ident })
-                        .collect()
-                }
-                hir::OpaqueTyOrigin::FnReturn { in_trait_or_impl, .. } => {
-                    if in_trait_or_impl.is_some()
-                        || self.tcx.features().lifetime_capture_rules_2024
-                        || span.at_least_rust_2024()
-                    {
-                        // return-position impl trait in trait was decided to capture all
-                        // in-scope lifetimes, which we collect for all opaques during resolution.
-                        self.resolver
-                            .take_extra_lifetime_params(opaque_ty_node_id)
-                            .into_iter()
-                            .map(|(ident, id, _)| Lifetime { id, ident })
-                            .collect()
-                    } else {
-                        // in fn return position, like the `fn test<'a>() -> impl Debug + 'a`
-                        // example, we only need to duplicate lifetimes that appear in the
-                        // bounds, since those are the only ones that are captured by the opaque.
-                        lifetime_collector::lifetimes_in_bounds(self.resolver, bounds)
-                    }
-                }
-                hir::OpaqueTyOrigin::AsyncFn { .. } => {
-                    unreachable!("should be using `lower_async_fn_ret_ty`")
-                }
+        // Whether this opaque always captures lifetimes in scope.
+        // Right now, this is all RPITIT and TAITs, and when `lifetime_capture_rules_2024`
+        // is enabled. We don't check the span of the edition, since this is done
+        // on a per-opaque basis to account for nested opaques.
+        let always_capture_in_scope = match origin {
+            _ if self.tcx.features().lifetime_capture_rules_2024 => true,
+            hir::OpaqueTyOrigin::TyAlias { .. } => true,
+            hir::OpaqueTyOrigin::FnReturn { in_trait_or_impl, .. } => in_trait_or_impl.is_some(),
+            hir::OpaqueTyOrigin::AsyncFn { .. } => {
+                unreachable!("should be using `lower_coroutine_fn_ret_ty`")
             }
         };
+        let captured_lifetimes_to_duplicate = lifetime_collector::lifetimes_for_opaque(
+            self.resolver,
+            always_capture_in_scope,
+            opaque_ty_node_id,
+            bounds,
+            span,
+        );
         debug!(?captured_lifetimes_to_duplicate);
 
         // Feature gate for RPITIT + use<..>
@@ -1920,7 +1883,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
 
         let captured_lifetimes = self
             .resolver
-            .take_extra_lifetime_params(opaque_ty_node_id)
+            .extra_lifetime_params(opaque_ty_node_id)
             .into_iter()
             .map(|(ident, id, _)| Lifetime { id, ident })
             .collect();

--- a/compiler/rustc_ast_lowering/src/lifetime_collector.rs
+++ b/compiler/rustc_ast_lowering/src/lifetime_collector.rs
@@ -1,5 +1,7 @@
 use rustc_ast::visit::{self, BoundKind, LifetimeCtxt, Visitor};
-use rustc_ast::{GenericBounds, Lifetime, NodeId, PathSegment, PolyTraitRef, Ty, TyKind};
+use rustc_ast::{
+    GenericBound, GenericBounds, Lifetime, NodeId, PathSegment, PolyTraitRef, Ty, TyKind,
+};
 use rustc_data_structures::fx::FxIndexSet;
 use rustc_hir::def::{DefKind, LifetimeRes, Res};
 use rustc_middle::span_bug;
@@ -10,14 +12,41 @@ use rustc_span::symbol::{Ident, kw};
 use super::ResolverAstLoweringExt;
 
 struct LifetimeCollectVisitor<'ast> {
-    resolver: &'ast ResolverAstLowering,
+    resolver: &'ast mut ResolverAstLowering,
+    always_capture_in_scope: bool,
     current_binders: Vec<NodeId>,
     collected_lifetimes: FxIndexSet<Lifetime>,
 }
 
 impl<'ast> LifetimeCollectVisitor<'ast> {
-    fn new(resolver: &'ast ResolverAstLowering) -> Self {
-        Self { resolver, current_binders: Vec::new(), collected_lifetimes: FxIndexSet::default() }
+    fn new(resolver: &'ast mut ResolverAstLowering, always_capture_in_scope: bool) -> Self {
+        Self {
+            resolver,
+            always_capture_in_scope,
+            current_binders: Vec::new(),
+            collected_lifetimes: FxIndexSet::default(),
+        }
+    }
+
+    fn visit_opaque(&mut self, opaque_ty_node_id: NodeId, bounds: &'ast GenericBounds, span: Span) {
+        // If we're edition 2024 or within a TAIT or RPITIT, *and* there is no
+        // `use<>` statement to override the default capture behavior, then
+        // capture all of the in-scope lifetimes.
+        if (self.always_capture_in_scope || span.at_least_rust_2024())
+            && bounds.iter().all(|bound| !matches!(bound, GenericBound::Use(..)))
+        {
+            for (ident, id, _) in self.resolver.extra_lifetime_params(opaque_ty_node_id) {
+                self.record_lifetime_use(Lifetime { id, ident });
+            }
+        }
+
+        // We also recurse on the bounds to make sure we capture all the lifetimes
+        // mentioned in the bounds. These may disagree with the `use<>` list, in which
+        // case we will error on these later. We will also recurse to visit any
+        // nested opaques, which may *implicitly* capture lifetimes.
+        for bound in bounds {
+            self.visit_param_bound(bound, BoundKind::Bound);
+        }
     }
 
     fn record_lifetime_use(&mut self, lifetime: Lifetime) {
@@ -99,6 +128,9 @@ impl<'ast> Visitor<'ast> for LifetimeCollectVisitor<'ast> {
                 self.record_elided_anchor(t.id, t.span);
                 visit::walk_ty(self, t);
             }
+            TyKind::ImplTrait(opaque_ty_node_id, bounds) => {
+                self.visit_opaque(*opaque_ty_node_id, bounds, t.span)
+            }
             _ => {
                 visit::walk_ty(self, t);
             }
@@ -106,13 +138,14 @@ impl<'ast> Visitor<'ast> for LifetimeCollectVisitor<'ast> {
     }
 }
 
-pub(crate) fn lifetimes_in_bounds(
-    resolver: &ResolverAstLowering,
+pub(crate) fn lifetimes_for_opaque(
+    resolver: &mut ResolverAstLowering,
+    always_capture_in_scope: bool,
+    opaque_ty_node_id: NodeId,
     bounds: &GenericBounds,
+    span: Span,
 ) -> FxIndexSet<Lifetime> {
-    let mut visitor = LifetimeCollectVisitor::new(resolver);
-    for bound in bounds {
-        visitor.visit_param_bound(bound, BoundKind::Bound);
-    }
+    let mut visitor = LifetimeCollectVisitor::new(resolver, always_capture_in_scope);
+    visitor.visit_opaque(opaque_ty_node_id, bounds, span);
     visitor.collected_lifetimes
 }

--- a/tests/ui/impl-trait/precise-capturing/capturing-implicit.rs
+++ b/tests/ui/impl-trait/precise-capturing/capturing-implicit.rs
@@ -1,0 +1,15 @@
+//@ edition: 2024
+//@ compile-flags: -Zunstable-options
+
+#![feature(rustc_attrs)]
+#![feature(type_alias_impl_trait)]
+#![rustc_variance_of_opaques]
+
+fn foo(x: &()) -> impl IntoIterator<Item = impl Sized> + use<> {
+    //~^ ERROR ['_: o]
+    //~| ERROR ['_: o]
+    //~| ERROR `impl Trait` captures lifetime parameter
+    [*x]
+}
+
+fn main() {}

--- a/tests/ui/impl-trait/precise-capturing/capturing-implicit.stderr
+++ b/tests/ui/impl-trait/precise-capturing/capturing-implicit.stderr
@@ -1,0 +1,22 @@
+error: `impl Trait` captures lifetime parameter, but it is not mentioned in `use<...>` precise captures list
+  --> $DIR/capturing-implicit.rs:8:11
+   |
+LL | fn foo(x: &()) -> impl IntoIterator<Item = impl Sized> + use<> {
+   |           ^       -------------------------------------------- lifetime captured due to being mentioned in the bounds of the `impl Trait`
+   |           |
+   |           this lifetime parameter is captured
+
+error: ['_: o]
+  --> $DIR/capturing-implicit.rs:8:19
+   |
+LL | fn foo(x: &()) -> impl IntoIterator<Item = impl Sized> + use<> {
+   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: ['_: o]
+  --> $DIR/capturing-implicit.rs:8:44
+   |
+LL | fn foo(x: &()) -> impl IntoIterator<Item = impl Sized> + use<> {
+   |                                            ^^^^^^^^^^
+
+error: aborting due to 3 previous errors
+

